### PR TITLE
Rebase OP-TEE linux driver from 3.4 to 3.14 based on kernel 5.13

### PR DESCRIPTION
**Pull request by xyzhao2018**
Signed-off-by: Shirley Zhao <shirley.zhao@intel.com>

### CODE REVIEW

File: Kconfig [hunk 1]
```diff
@@ -2,7 +2,7 @@
 # Generic Trusted Execution Environment Configuration
 config TEE
 	tristate "Trusted Execution Environment support"
-	depends on HAVE_ARM_SMCCC || COMPILE_TEST || CPU_SUP_AMD
```
Comment (jingdlu) [inline Kconfig:5]: Don't need to remove CPU_SUP_AMD.

File: Kconfig [hunk 2]
```diff
@@ -16,7 +16,6 @@ if TEE
 menu "TEE drivers"
 
 source "drivers/tee/optee/Kconfig"
-source "drivers/tee/amdtee/Kconfig"
```
Comment (jingdlu) [inline Kconfig]: Why do we need to remove this amdtee line?

File: optee/Kconfig [hunk 3]
```diff
@@ -2,7 +2,7 @@
 # OP-TEE Trusted Execution Environment Configuration
 config OPTEE
 	tristate "OP-TEE"
-	depends on HAVE_ARM_SMCCC
+	depends on HAVE_ARM_SMCCC || X8_64
```
Comment (jingdlu) [inline optee/Kconfig]: X86_64?

File: optee/core.c [hunk 4]
```diff
@@ -684,6 +741,7 @@ static int optee_probe(struct platform_device *pdev)
 	mutex_init(&optee->call_queue.mutex);
 	INIT_LIST_HEAD(&optee->call_queue.waiters);
 	optee_wait_queue_init(&optee->wait_queue);
+
```
Comment (jingdlu) [inline optee/core.c]: Don't need to add this blank line.

File: optee/core.c [hunk 5]
```diff
@@ -711,6 +769,7 @@ static int optee_probe(struct platform_device *pdev)
 		 * devices hasn't been registered with
 		 * tee_device_register() yet.
 		 */
+
```
Comment (jingdlu) [inline optee/core.c]: Don't need to add this blank line.

File: optee/core.c [hunk 6]
```diff
@@ -723,20 +782,71 @@ static int optee_probe(struct platform_device *pdev)
...
 MODULE_DEVICE_TABLE(of, optee_dt_match);
 
+static struct device_node optee_dev_node = {
+	.name = "optee-tz",
+};
+
+void optee_dev_release(struct device *dev)
+{
+	return;
+}
```
Comment (jingdlu) [inline optee/core.c:797]: Please put all of these into #if defined(CONFIG_X86_64)

File: optee/core.c [hunk 7]
```diff
@@ -723,20 +782,71 @@ static int optee_probe(struct platform_device *pdev)
...
+static void __exit optee_drv_exit(void)
+{
+	platform_driver_unregister(&optee_driver);
+
+#ifndef CONFIG_OF_EARLY_FLATTREE
+		platform_device_unregister(&optee_platform_dev);
+#endif
+}
+
+module_init(optee_drv_init);
```
Comment (jingdlu) [inline optee/core.c:848]: We still can't use module_platform_driver(optee_driver)?

File: optee/core.c [hunk 8]
```diff
@@ -723,20 +782,71 @@ static int optee_probe(struct platform_device *pdev)
...
-module_platform_driver(optee_driver);
+
+static int __init optee_drv_init(void)
+{
+	int ret;
+
+#ifndef CONFIG_OF_EARLY_FLATTREE
+	ret = platform_add_devices(optee_devices, ARRAY_SIZE(optee_devices));
+	if (ret) {
+		printk(KERN_ERR "%s(): platform_add_devices() failed, ret %d\n", __func__, ret);
```
Comment (jingdlu) [inline optee/core.c]: Use pr_err.

File: optee/core.c (hunk 7 above)
Comment (xyzhao2018) [reply optee/core.c:848]: The driver has been registered/unregistered in device initialize/remove process.

**Review by GangSecurity — APPROVED** [review]

**Review by jingdlu — APPROVED** [review]

**Pull request merged**